src/AnalysisEpic.cxx

```diff
@@ -0,0 +1,424 @@
+      
+    for(int itrack=0; itrack < tracks_PDG.GetSize(); itrack++){
+
+      int pid_ = tracks_PDG[itrack];
+      double px_ = tracks_p_x[itrack];
+      double py_ = tracks_p_y[itrack];
+      double pz_ = tracks_p_z[itrack];
+      double e_ = tracks_e[itrack];
+      double m_ = sqrt(e_*e_-px_*px_+py_*py_+pz_*pz_);
+      
+      // Add to trackpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+
+      /*
+	Read through Associations to match particles
+	By default, we assume no association, so mcID --> -1
+	assoc_recID --> itrack (index of the RecoParticle)
+	assoc_simID --> imc (index of the MCParticle)
+      */
+
+
+      part.mcID=-1;
+      for(int iassoc = 0 ; iassoc < assoc_simID.GetSize() ; iassoc++){
+	int idx_recID = assoc_recID[iassoc]; 
+	int idx_simID = assoc_simID[iassoc];
+	if(itrack==idx_recID){ // This track has an association
+	  part.mcID=idx_simID;
+	  break; // Only one association per particle
+	}
+      }
+
+      trackpart.push_back(part);
+      trackidmap.insert({itrack,part.mcID}); 
+      
+    }
+
+
+
+    /*
+      With the GeneratedParticles, MCParticles, and ReconstructedParticles filled,
+      we can begin to search for the beam particles and hadronic final state (HFS)
+      This is done for both the Truth and Reconstructed Particles
+    */
+
+    /*
+      Loop over MCParticles
+    */
+
+    for(auto mcpart_: mcpart){
+
+      int imc = mcpart_.mcID;
+      /* Beam particles have a MCParticles.generatorStatus of 4 */
+      int genStat_ = mcpart_genStat[imc];
+      if(mcpart_.pid==11 && genStat_ == 4){
+	foundBeamElectron=true;
+	kinTrue->vecEleBeam = mcpart_.vecPart;
+      }
+      else if(mcpart_.pid==2212 && genStat_ == 4){
+	foundBeamIon=true;
+	kinTrue->vecIonBeam = mcpart_.vecPart;
+      }
+      else if(genStat_==4){
+	cout << "Warning...unknown beam particle with generatorStatus == 4 found...Continuing..." << endl;
+      }
+
+      /* Assume the scattered electron is the pid==11 final state particle with the most energy */
+      if(mcpart_.pid==11 && genStat_ == 1 && mcpart_.vecPart.P() > maxP)
+	{
+	  maxP=mcpart_.vecPart.P();
+	  kinTrue->vecElectron = mcpart_.vecPart;
+	  genEleID = mcpart_.mcID; 
+	}
+
+      /*
+	Only append MCParticles to the HFS if they are matched with a GeneratedParticle
+       */
+      
+      else if(genStat_ == 1 && mcidmap[mcpart_.mcID]>-1){ 
+	kinTrue->AddToHFS(mcpart_.vecPart);
+      }
+    
+    }
+
+    //check beam finding
+    if(!foundBeamElectron || !foundBeamIon) { numNoBeam++; continue;};
+
+    /*
+      Loop over RecoParticles
+    */
+
+    int itrack = 0;
+    bool recEleFound=false;
+    for(auto trackpart_ : trackpart){
+      // Skip if there is no matching MCParticle
+      if(trackidmap[itrack]==-1) continue;
```

Comment:
does `itrack` still need to iterate if this `continue` is called?

Reply:
Yes it does, fixed

Reply:
Yes it does, fixed